cordapp-cpk/src/main/kotlin/net/corda/plugins/cpb/CpbTask.kt

```diff
@@ -4,27 +4,42 @@ import net.corda.plugins.cpk.CORDAPP_TASK_GROUP
 import net.corda.plugins.cpk.CORDA_CPK_TYPE
 import net.corda.plugins.cpk.CPK_FILE_EXTENSION
 import org.gradle.api.file.DuplicatesStrategy
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
 import org.gradle.api.tasks.AbstractCopyTask
+import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.bundling.Jar
 import org.gradle.api.tasks.bundling.ZipEntryCompression
 import java.nio.file.Files
 import java.util.jar.JarInputStream
 import java.util.zip.ZipInputStream
+import javax.inject.Inject
 
-open class CpbTask : Jar() {
+open class CpbTask @Inject constructor(objects: ObjectFactory) : Jar() {
 
     companion object {
         private const val CPB_ARTIFACT_CLASSIFIER = "package"
         const val CPB_FILE_EXTENSION = "cpb"
         private const val CPK_FILE_SUFFIX = ".$CPK_FILE_EXTENSION"
         private const val JAR_FILE_SUFFIX = ".jar"
         private val EXCLUDED_CPK_TYPES = setOf("corda-api")
+        const val CPB_NAME_ATTRIBUTE = "Corda-CPB-Name"
+        const val CPB_VERSION_ATTRIBUTE = "Corda-CPB-Version"
     }
 
+
+    @get:Input
+    val cpbName : Property<String>
+
+    @get:Input
+    val cpbVersion : Property<String>
```

Comment:
I just didn't want to tightly couple the name of the file to the name written inside the manifest... does it make any sense? Idk... probably not as you can always rename the file if, for any reason (and I cannot think of any reason), you are unhappy with it

Reply:
Give people a choice of which "version" property to set, and I _guarantee_ that at least some will choose the wrong one.

Come to think of it, we already have two versions anyway:
- The Gradle/Maven version.
- The OSGi bundle version.

I would actually expect `CPB_VERSION_ATTRIBUTE` to be set to one of those two, rather than inventing a third kind of version. Maybe check with @adelel1 which one we really want here?

Reply:
a cpb file can contain multiple CPKs so I don't think it makes sense to use either of those

Reply:
It is basically not tight to a single CPK, even though it is probably sensible to have a default that matches the name and version of the current project. But I think that in principle you should be able to assemble a CPB that contains 2 independent CPKs, and you cannot tell which one is the *main*

Reply:
Fair enough about the bundle version. In which case, I just cannot imagine a use-case where "CPB version" is expected to mean anything different from "Gradle/Maven" version. Consider:
> Here is version 2 of my CPB: `bongo-3.4.76-SNAPSHOT-package.cpb`

Reply:
yeah, I agree... as you said it is probably better not to build to many levers because the user will eventually pull the one with the big **"DO NOT PULL"** sign over it

Reply:
> yeah, I agree... as you said it is probably better not to build to many levers because the user will eventually pull the one with the big **"DO NOT PULL"** sign over it

In which case it's probably better to remove `cpbVersion`, and just store `archiveVersion` in the CPB manifest.

What is the use-case for the `CPB-Name` property?

Reply:
as part of the unique identifier for the cpb inside Corda, atm we use the archive hash but we would like to move towards a tuple like `(name, version)` (still unclear whether signature should be part of it)

Reply:
> as part of the unique identifier for the cpb inside Corda, atm we use the archive hash but we would like to move towards a tuple like `(name, version)` (still unclear whether signature should be part of it)

The question is "What does a CPB name look like?", or more specifically "Is it meant to be human-friendly or machine-friendly?".

For example:
```gradle
cpb {
    cpbName = 'FooBar-SDK-1'
}
```
or
```gradle
cpb {
    cpbName = 'Basil, the Ever-Hungry slayer of wildebeast!'
}
```
I suppose it's a question of whether some characters (such as spaces, colons etc) are "allowed" or not.

Reply:
I don't know tbh... for now I think it will be only used internally by software with little exposure to the user but this might change in the future

Reply:
space in manifest attribute are not an issue though

Reply:
> space in manifest attribute are not an issue though

You can have file names with spaces in too, although I don't recommend it. Characters like quotes and punctuation are more problematic, particularly on Windows. But the point of using `archiveBaseName` is that it's a file name _first_ and a CPB name _second_, and Gradle will complain if it's not a valid file name.